### PR TITLE
[IOTDB-890] SDT implementation

### PR DESCRIPTION
**Pull request by haimeiguo**
implements SDT in chunk writer. data gets discarded when "flush" in memory before going to disk.

### CODE REVIEW

**Comment by HTHou** [comment]
The link of the design doc is here. :) https://cwiki.apache.org/confluence/display/IOTDB/SDT+encoding

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4 [hunk 1]
```diff
@@ -641,6 +640,7 @@ dateFormat
 constant
     : dateExpression
     | NaN
+    | SDT
```
Comment (mychaow) [inline antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4:643]: Maybe you need to add some parameter name of sdt?
And Please add some example in the docs.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 2]
```diff
@@ -97,18 +109,93 @@ public ChunkWriterImpl(MeasurementSchema schema) {
...
+        //check if is sdt encoding
+        if (entry.getKey().toLowerCase().equals("loss") && entry.getValue().toLowerCase()
+            .equals("sdt")) {
+          isSdtEncoding = true;
+          sdtEncoder = new SdtEncoder();
+        }
+
+        //set compression deviation
+        else if (isSdtEncoding &&
+            entry.getKey().toLowerCase().equals("cd")) {
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:134]: cd is too simple

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 3]
```diff
@@ -155,6 +266,12 @@ public void write(long[] timestamps, boolean[] values, int batchSize) {
 
   @Override
   public void write(long[] timestamps, float[] values, int batchSize) {
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:268]: so, how about double[] ?

**Review by mychaow — CHANGES_REQUESTED** [review]

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 4]
```diff
@@ -97,18 +109,93 @@ public ChunkWriterImpl(MeasurementSchema schema) {
...
+  }
+
+  public ChunkWriterImpl(MeasurementSchema schema, boolean isMerging) {
+    this(schema);
+    this.isMerging = isMerging;
+  }
+
+  private void checkSdtEncoding() {
+    if (measurementSchema.getProps() != null && !isMerging) {
+      for (Map.Entry<String, String> entry : measurementSchema.getProps().entrySet()) {
```
Comment (JackieTien97) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:124]: No need to iterate the map, and also there is no guarantee that "loss" property will be got before other sdt parameters.

You can directly use map.contains and map.get() to get whatever property you want.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
Good job!
I've got one question, it seems that your sdt algorithm is not in line with your referenced paper in your design document(https://sol.sbc.org.br/index.php/sbesc_estendido/article/download/8650/8551#:~:text=The%20Swing%20Door%20Trending%20(SDT,process%20information%20systems%20(PIMs)).
The paper calculate the FV of the next segment but you just use the endPoint of the last segment.

And anyway, I also think that SdtEncoder should use primitive type in java instead of the Long, Double, Float in case of unnecessary unbox and box operation. You can see how we implement the BatchData to deal with different types data.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java (hunk 4 above)
Comment (haimeiguo) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:124]: will do. thank you. 
For "The paper calculate the FV of the next segment but you just use the endPoint of the last segment.", what do you mean by that? When ending segment0 and starting segment1, I use segment0's endpoint as segment1's startpoint. How do you calculate segment1's first value..?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+      lastStoredInt = lastReadInt;
+      upperDoor = (value - lastStoredInt.right - compDeviation) / (time - lastStoredInt.left);
+      lowerDoor = (value - lastStoredInt.right + compDeviation) / (time - lastStoredInt.left);
+      lastReadInt = new Pair<>(time, value);
+      return true;
+    }
+
+    lastReadInt = new Pair<>(time, value);
+    return false;
+  }
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java:247]: need a blank line

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    }
+
+    // if current point to the last stored point's time distance is larger than compMax,
+    // will reset two doors, and store current point;
+    if (time - lastStoredDouble.left >= compMax) {
+      reset(time, value);
+      return true;
+    }
+
+    double curUpperSlope = (value - lastStoredDouble.right - compDeviation) / (time - lastStoredDouble.left);
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java:267]: maybe have divide zero exception

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+public class SdtEncoder {
+
+  /**
+   * the last read pair <time, value></>
+   * if upperDoor >= lowerDoor meaning out of compDeviation range, will store lastReadPair
+   */
+  private Pair<Long, Double> lastReadDouble;
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java:34]: use primitive data type, this is also has box cost.

**Review by mychaow — CHANGES_REQUESTED** [review]

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4 [hunk 8]
```diff
@@ -1272,6 +1268,8 @@ UNDERLINE : '_';
 
 NaN : 'NaN';
 
+SDT : 'SDT';
```
Comment (Genius-pig) [inline antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4:1271]: Why write like this. Does SDT didn't need case insensitive?

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4 (hunk 1 above)
Comment (Genius-pig) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4:643]: if SDT is a constant, I don't think it needs a keyword.

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4 (hunk 8 above)
Comment (haimeiguo) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/strategy/SqlBase.g4]: removed. thank you!

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 9]
```diff
@@ -48,28 +48,41 @@
 
   public static void main(String[] args)
       throws IoTDBConnectionException, StatementExecutionException {
-    session = new Session("127.0.0.1", 6667, "root", "root");
+    session = new Session("172.16.48.4", 58890, "root", "root");
```
Comment (qiaojialin) [inline example/session/src/main/java/org/apache/iotdb/SessionExample.java:51]: Hi, please recover this class.

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 10]
```diff
@@ -48,28 +48,41 @@
...
+   // session.executeNonQueryStatement("delete storage group root");
+    /*
     try {
       session.setStorageGroup("root.sg1");
     } catch (StatementExecutionException e) {
       if (e.getStatusCode() != TSStatusCode.PATH_ALREADY_EXIST_ERROR.getStatusCode())
         throw e;
     }
 
-    createTimeseries();
```
Comment (qiaojialin) [inline example/session/src/main/java/org/apache/iotdb/SessionExample.java:61]: could the SDT be set using Session nosql API?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 11]
```diff
@@ -119,14 +194,26 @@ public void write(long time, boolean value) {
...
   }
 
   @Override
   public void write(long time, double value) {
-    pageWriter.write(time, value);
-    checkPageSizeAndMayOpenANewPage();
+    if (!isSdtEncoding || sdtEncoder.encodeDouble(time, value)) {
+      if (isSdtEncoding) {
+        time = sdtEncoder.getTime();
+        value = (double) sdtEncoder.getValue();
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:212]: could we getDoubleValue()?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+ */
+
+package org.apache.iotdb.tsfile.encoding.encoder;
+
+import java.nio.FloatBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+
+public class SdtEncoder {
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SdtEncoder.java:27]:
```suggestion
public class SDTEncoder {
```

**Comment by qiaojialin** [comment]
Besides, there are many code smells, try to fix them ~

**Review by JackieTien97 — APPROVED** [review]

File: tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SDTEncoder.java [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+      lastStoredDouble = value;
+      return true;
+    }
+    return false;
+  }
+
+  private void reset() {
+    upperDoor = Integer.MIN_VALUE;
+    lowerDoor = Integer.MAX_VALUE;
+    isFirstValue = true;
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/SDTEncoder.java:464]: false?

**Review by JackieTien97 — APPROVED** [review]

**Review by qiaojialin — APPROVED** [review]
LGTM~

**Review by mychaow — APPROVED** [review]
LGTM

**Pull request merged**